pyphare/pyphare/core/operators.py

```diff
@@ -0,0 +1,141 @@
+import numpy as np
+
+from pyphare.pharesee.hierarchy import ScalarField, VectorField  # TensorField
+from pyphare.pharesee.hierarchy.hierarchy_utils import compute_hier_from
+from pyphare.pharesee.hierarchy.hierarchy_utils import rename
+
+
+def _compute_dot_product(patch_datas, **kwargs):
+    ref_name = next(iter(patch_datas.keys()))
+
+    dset = (
+        patch_datas["left_x"][:] * patch_datas["right_x"][:]
+        + patch_datas["left_y"][:] * patch_datas["right_y"][:]
+        + patch_datas["left_z"][:] * patch_datas["right_z"][:]
+    )
+
+    return (
+        {"name": "value", "data": dset, "centering": patch_datas[ref_name].centerings},
+    )
+
+
+def _compute_sqrt(patch_datas, **kwargs):
+    ref_name = next(iter(patch_datas.keys()))
+
+    dset = np.sqrt(patch_datas["value"][:])
+
+    return (
+        {"name": "value", "data": dset, "centering": patch_datas[ref_name].centerings},
+    )
+
+
+def _compute_cross_product(patch_datas, **kwargs):
+    ref_name = next(iter(patch_datas.keys()))
+
+    dset_x = (
+        patch_datas["left_y"][:] * patch_datas["right_z"][:]
+        - patch_datas["left_z"][:] * patch_datas["right_y"][:]
+    )
+    dset_y = (
+        patch_datas["left_z"][:] * patch_datas["right_x"][:]
+        - patch_datas["left_x"][:] * patch_datas["right_z"][:]
+    )
+    dset_z = (
+        patch_datas["left_x"][:] * patch_datas["right_y"][:]
+        - patch_datas["left_y"][:] * patch_datas["right_x"][:]
+    )
+
+    return (
+        {"name": "x", "data": dset_x, "centering": patch_datas[ref_name].centerings},
+        {"name": "y", "data": dset_y, "centering": patch_datas[ref_name].centerings},
+        {"name": "z", "data": dset_z, "centering": patch_datas[ref_name].centerings},
+    )
+
+
+def _compute_grad(patch_data, **kwargs):
+    ndim = patch_data["value"].box.ndim
+    nb_ghosts = kwargs["nb_ghosts"]
+    ds = patch_data["value"].dataset
+
+    ds_shape = list(ds.shape)
+
+    ds_x = np.full(ds_shape, np.nan)
+    ds_y = np.full(ds_shape, np.nan)
+    ds_z = np.full(ds_shape, np.nan)
+
+    grad_ds = np.gradient(ds)
+    select = tuple([slice(nb_ghosts, -nb_ghosts) for _ in range(ndim)])
+    if ndim == 2:
+        ds_x[select] = np.asarray(grad_ds[0][select])
+        ds_y[select] = np.asarray(grad_ds[1][select])
+        ds_z[select].fill(0.0)  # TODO at 2D, gradient is null in z dir
```

Comment:
todo, or torm?

Reply:
I'll leave the comment I think @rochSmets meant todo for 3D probably